packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,514 @@
+    // TODO check oob state is await-respons
+    // TODO check there is no connection record for particular oob record
+
+    const { did, mediatorId } = routing ? routing : outOfBandRecord
+    if (!did) {
+      throw new AriesFrameworkError('Out-of-band record does not have did attribute.')
+    }
+
+    const { message } = messageContext
+
+    // Check corresponding invitation ID is the request's ~thread.pthid
+    // TODO Maybe we can do it in handler, but that actually does not make sense because we try to find oob by parent thread ID there.
+    if (!message.thread?.parentThreadId || message.thread?.parentThreadId !== outOfBandRecord.getTags().messageId) {
+      throw new DidExchangeProblemReportError('Missing reference to invitation.', {
+        problemCode: DidExchangeProblemReportReason.RequestNotAccepted,
+      })
+    }
+
+    // If the responder wishes to continue the exchange, they will persist the received information in their wallet.
+
+    if (!message.did.startsWith('did:peer:')) {
+      throw new DidExchangeProblemReportError(
+        `Message contains unsupported did ${message.did}. Supported dids are [did:peer]`,
+        {
+          problemCode: DidExchangeProblemReportReason.RequestNotAccepted,
+        }
+      )
+    }
+    const peerDid = DidPeer.fromDid(message.did)
+    if (peerDid.numAlgo !== PeerDidNumAlgo.GenesisDoc) {
+      throw new DidExchangeProblemReportError(
+        `Unsupported numalgo ${peerDid.numAlgo}. Supported numalgos are [${PeerDidNumAlgo.GenesisDoc}]`,
+        {
+          problemCode: DidExchangeProblemReportReason.RequestNotAccepted,
+        }
+      )
+    }
+
+    const didDocument = await this.extractDidDocument(message)
+    const didRecord = new DidRecord({
+      id: message.did,
+      role: DidDocumentRole.Received,
+      // It is important to take the did document from the PeerDid class
+      // as it will have the id property
+      didDocument,
+      tags: {
+        // We need to save the recipientKeys, so we can find the associated did
+        // of a key when we receive a message from another connection.
+        recipientKeys: didDocument.recipientKeys.map(verkeyToDidKey),
+      },
+    })
+
+    this.logger.debug('Saving DID record', {
+      id: didRecord.id,
+      role: didRecord.role,
+      tags: didRecord.getTags(),
+      didDocument: 'omitted...',
+    })
+
+    await this.didRepository.save(didRecord)
+
+    const connectionRecord = await this.connectionService.createConnection({
+      protocol: HandshakeProtocol.DidExchange,
+      role: DidExchangeRole.Responder,
+      state: DidExchangeState.RequestReceived,
+      multiUseInvitation: false,
+      did,
+      mediatorId,
+      autoAcceptConnection: outOfBandRecord.autoAcceptConnection,
+      outOfBandId: outOfBandRecord.id,
+    })
+    connectionRecord.theirDid = message.did
+    connectionRecord.theirLabel = message.label
+    connectionRecord.threadId = message.threadId || message.id
```

Comment:
Vau, I would not expect that at all. Is it the case for all DIDcomm protocols? What about cases where you want to be sure that `.threadId`is present regardless `.id` or that it should be different?

Reply:
Then you can check the `thread.threadId`, but according to the RFC the thread is either the thread.threadId or otherwise it's implicitly the the message `@id` 🤷